### PR TITLE
[Gh 904] Central Catalog Support 

### PR DESCRIPTION
**Pull request by TejasRGitHub**
### Feature or Bugfix
- Feature

### Detail

PR containing all the code raised in  PR - https://github.com/data-dot-all/dataall/pull/905 + Unit Tests + Addressing comments raised on that PR. Copy pasting details from PR - 

Detect if the source database is a resource link
If it is a resource link, check that the catalog account has been onboarded to data.all
Check for the presence of owner_account_id tag on the database
The tag needs to exist and the value has to match the account id of the share approver

Credits - @blitzmohit 

## Testing 

Running Unit tests - ✅ 
Testing on AWS Deployed data.all instance with the Original PR -  ✅ 
Sanity testing after addressing comments -  **[EDIT]** ✅  ( Testing done ) 

### Relates
- https://github.com/data-dot-all/dataall/issues/904

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)? No
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization? No
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features? No
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users? Yes
  - Have you used the least-privilege principle? How? Yes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by TejasRGitHub** [comment]
Adding the same instructions as mentioned by @blitzmohit  on the previous PR for easy access - 

Expected flow is as follows:

Data.all Onboarding:

Producer_team is onboarded to data.all with their account/env set up.
Consumer_team is onboarded to data.all with their account/env set up.
Data_catalog_team is onboarded to data.all with their account/env set up. (Note: No additional privileges are required on the catalog account)
Glue DB Creation:

Producer_team requests data_catalog_team to create a Glue database (e.g., producer_team_db1).
Data_catalog_team creates the Glue DB in their account and shares it back with write/create table permissions to producer_team.
Data_catalog_team adds a tag on producer_team_db1 with the key "owner_account_id" and the value as producer_team’s AWS account ID (account_id: 11111).
Data.all Dataset & Share Creation:

Producer_team imports a new dataset (dataset1) with the S3 bucket & producer_team_db1 Glue DB.
Consumer_team creates a share request for specific Glue tables & S3 bucket access in the producer_team dataset.
Data.all Share Provisioning:

Producer_team approves the share request.
Share is provisioned if the following conditions are met:
Data.all checks the Glue DB metadata to get catalog account id
Check the catalog account is onboarded to data.all (i.e., data.all can assume the pivotRole in the catalog account).
The "owner_account_id" tag exists on producer_team_db1 Glue DB.
The tag value matches with producer_team account ID (i.e. {owner_account_id=11111})
Note that is it is intended to be used with the S3 bucket access as sharing the table from the catalog account would not provide that.

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:91]: We need to resolve the conflicts between this PR and the [one for simplifying](https://github.com/data-dot-all/dataall/pull/1016) database names. That PR removes the 2 processors and uses a single one for same account and cross account.

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:263]: In general I think data_sharing_service should include minimal "sharing actions", instead the processors should be the ones executing the particular sharing mechanisms for approve and revoke. 
- `service` initializes the `processors` and calls `approve` or `revoke`
- `processor` defines `approve` or `revoke` workflow using `manager` functions
- `manager` defines actions that grant/revoke/configure the share. Typically use a AWS service
- `x_service_client` defines the boto3 calls as abstract as possible

Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:263]: Following these principles, we should:
- Ultra minor: `def get_source_catalog(self):` rename to  `def get_catalog(self):` and remove references to source account because it depends on the GlueClient init values
- with the changes in the [PR](https://github.com/data-dot-all/dataall/pull/1016) for simplify db names we do not need to define the source account in the processor initialization. I suggest we do not pass the `catalog_details` from the sharing_service. Instead we should include this block of code directly in `lf_share_manager`

File: backend/dataall/modules/dataset_sharing/db/share_object_models.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/db/share_object_models.py:63]: I am not sure we will need this class if we work directly in the `lf_share_manager`. If it stays, I suggest we move it out of the `db` models and define it where it is used. I would also rename it to `GlueDatabase`

**Review by dlpzx — CHANGES_REQUESTED** [review]
Some changes needed left in comments

**Comment by TejasRGitHub** [comment]
Hi @dlpzx , I have updated PR by merging into the changes made by you for Simplifying the Lake formation. Thanks for the refactoring, the code really looks very clean and is easy to understand. 

Pending task - 

- [ ] Raising another PR to update the readme to share more on how the catalog model works 
- [ ] Resolving failing Unit test and adding tests for the new introduced function

I have performed the following tests

## Testing

1. Existing normal share ( with a producer and a consumer ) 

- Revoking Share ✅ 
- Approving Share ✅ ( By Revoking one table from share and then again approving that table and checking if the existing shared db is used ) 

2. Creating a new share on a dataset 

- Approving share ✅  ( Able to see a shared db with only the "_shared" )
- Revoking share ✅  ( The "_shared" db is removed and no table is seen when checking from Athena for that role)

3. Existing share on catalog database 

- Revoking a share ( TODO )
- Approving share ( with additional tables or the one which was revoked ) (TODO)

4. Catalog account share 

- Checking share with a consumer account ( which is not a producer and catalog account ) 
    - Adding 1 Table for share ✅  ( Able to see the "_shared" db created in consumer account )
    - Adding another Table for share ✅  ( The same "_shared" db is used and share is successful)
    - Revoking 1 Table ✅  ( "_shared" db is present and now able to see the revoked table ) 
    - Revoking all other tables ✅  ( "_shared" is not present ) 
    
- Checking share with the producer account 
    - Creating the share of table with a Team ( Role ) which is present in producer env ✅ 
    - Revoking the share of table with that same team ✅ 
    
- Checking share with the catalog account itself 
    - Creating the share of table with a Team ( Role ) which is present in producer env ✅ 
    - Revoking the share of table with that same team ✅

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (TejasRGitHub) [reply backend/dataall/modules/dataset_sharing/services/data_sharing_service.py]: Hi @dlpzx , I have refactored the code as per your recommendations and the code structure that is followed

File: backend/dataall/modules/dataset_sharing/db/share_object_models.py
Comment (TejasRGitHub) [reply backend/dataall/modules/dataset_sharing/db/share_object_models.py]: Removed this class as not needed anymore

File: backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:167]: `check_catalog_account_exists_and_update_processor` will return True even when `catalog_dict` = `None`. Which means that this `if` will always be True and lines 168,169 will never be executed (I think)

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:433]: The function `handle_share_failure_for_all_tables` is used both in the approve and in the revoke processes but it calls handle_revoke_failure for both processes

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:454]: Can we split `check_catalog_account_exists_and_update_processor` into 2 functions each with one task: `check_database_catalog_account` that gets the database source catalog and calls `_verify_catalog_ownership` and then depending on the output of this fucntion we can call `update_processor_source_account`

**Review by dlpzx — COMMENTED** [review]
Left some comments. Going back to the root issue, I am thinking that maybe we could add the `check_catalog_account_exists_and_update_processor` logic as part of the initialization of the `LFShareManager`. 

With this PR, when we approve a share request:
1. We initialize the `ProcessLakeFormationShare(LFShareManager)` with "source" variables of the source_environment
2. Then we call `process_approved_shares`
3. Where we`check_catalog_account` and then `update_processor` variable (re-writing the "source" values defined in the __init__)
4. If `check_catalog_account` raises exception we `handle_share_failure_for_all_tables`

As an alternative we could:
1. We initialize the `ProcessLakeFormationShare(LFShareManager)`. In the __init__ we `check_catalog_account` and we directly define the "source" variables with their final values. Similar to how we define the `build_db_shared`. 

I think the main issue is the exception handling for the catalog when the account is not onboarded to data.all and it is not properly tagged. We could store that the catalog is invalid for example setting the `source_account_id` as None and then manage the alarms in `process_approved_shares`


We can meet over a call to talk about it, but what do you think? I just thought that we could make it more human-logical and avoid re-writing init values

File: backend/dataall/base/aws/sts.py
Comment (noah-paige) [inline backend/dataall/base/aws/sts.py:365]: will this ever return None or just raise an Exception if no session is created?

File: backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py
Comment (noah-paige) [inline backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:70]: minor - let's just return false here

**Comment by noah-paige** [comment]
> Left some comments. Going back to the root issue, I am thinking that maybe we could add the `check_catalog_account_exists_and_update_processor` logic as part of the initialization of the `LFShareManager`.
> 
> With this PR, when we approve a share request:
> 
>     1. We initialize the `ProcessLakeFormationShare(LFShareManager)` with "source" variables of the source_environment
> 
>     2. Then we call `process_approved_shares`
> 
>     3. Where we`check_catalog_account` and then `update_processor` variable (re-writing the "source" values defined in the **init**)
> 
>     4. If `check_catalog_account` raises exception we `handle_share_failure_for_all_tables`
> 
> 
> As an alternative we could:
> 
>     1. We initialize the `ProcessLakeFormationShare(LFShareManager)`. In the **init** we `check_catalog_account` and we directly define the "source" variables with their final values. Similar to how we define the `build_db_shared`.
> 
> 
> I think the main issue is the exception handling for the catalog when the account is not onboarded to data.all and it is not properly tagged. We could store that the catalog is invalid for example setting the `source_account_id` as None and then manage the alarms in `process_approved_shares`
> 
> We can meet over a call to talk about it, but what do you think? I just thought that we could make it more human-logical and avoid re-writing init values

Agreed with this approach - think best if we avoid setting and re-setting instance variables and would be cleaner to just handle at `init()`

File: backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py
Comment (TejasRGitHub) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:167]: Hi @dlpzx , Thanks for raising this. Yes, the if condition will never be executed. I had added this when the outer try catch was not present( See line 166 , 170 ) . I will correct it and send the updated PR

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (TejasRGitHub) [reply backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:433]: Will update this . Thanks @dlpzx

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (TejasRGitHub) [reply backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:435]: Agreed. Will make the change accordingly

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (TejasRGitHub) [reply backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:454]: Yes. I could do that.

File: backend/dataall/base/aws/sts.py
Comment (TejasRGitHub) [reply backend/dataall/base/aws/sts.py:365]: It should just return False if no session is created

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (TejasRGitHub) [inline backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:107]: Returning '', True which will happen when the source database name is not set properly  which is likely if there are errors in setting catalog account details. The validation to check if the source database name , source account id, source region is done in the `process_revoked_shares` and `process_approved_shares` function

**Comment by TejasRGitHub** [comment]
Updated the PR after refactoring as per review comment. @dlpzx , @noah-paige , please take a look and review it. Thanks!

### Testing 

1. Share with normal producer consumer dataset with table and S3 bucket ✅ 
2. Share Revoke with normal producer and consumer of dataset with Table and S3Bucket ✅ 
3. Share with Central Catalog dataset with the Consumer ✅ 
4. Share Revoke Central Catalog dataset with the Consumer ✅ 
5. Share with Central Catalog dataset with the Producer ✅ 
6. Share  Revoke with Central Catalog dataset with the Producer ✅ 
7. Share with Central Catalog dataset with the  Central Catalog Account’s role ✅ 
8. Share Revoke with Central Catalog dataset with the  Central Catalog Account’s role ✅ 

For all testing scenario , checked if the tables are showing up in Athena for the role on which they are shared . Also confirmed if when the share is revoked , the _shared table is removed and the permissions are not present.

File: backend/dataall/base/aws/sts.py
Comment (dlpzx) [reply backend/dataall/base/aws/sts.py:365]: What happens when the session goes into a ClientError exception (`dataall/base/aws/sts.py:64`?

File: backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:72]: It is minor, but this is still present in the last commited version

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:66]: This function is a bit confusing, when you say "present" you mean that there is a separate catalog_account different from the source_environment.AwsAccount and that the catalog exists and is tagged right?

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:510]: Very minor: instead of checking the `if catalog_account_id != self.source_environment.AwsAccountId:` inside `_verify_catalog_ownership`, we could add the condition in line 508: `if catalog_dict is not None and catalog_account_id != self.source_environment.AwsAccountId:`

File: backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:85]: Line 83 return `False` directly

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:524]: This function is a bit a duplication of `check_catalog_account_exists_and_verify`.We could have a single function:
```
def verify_and_get_catalog_account_details(self):
        try:
            catalog_dict = GlueClient(
                account_id=self.source_environment.AwsAccountId,
                region=self.source_environment.region,
                database=self.dataset.GlueDatabaseName,
            ).get_source_catalog()
            if catalog_dict is None:
                    logger.info(
                    f'No Catalog information found for dataset - {self.dataset.name} containing database - {self.dataset.GlueDatabaseName}')
            elif catalog_dict.get('account_id') != self.source_environment.AwsAccountId:
                # Verify the ownership of dataset by checking if pivot role is assumable and ownership tag is present
                self._verify_catalog_ownership(catalog_dict.get('account_id'), catalog_dict.get('region'),
                                               catalog_dict.get('database_name'))
            return catalog_dict.get('account_id', None), catalog_dict.get('region', None),
                                               catalog_dict.get('database_name', None)
        except Exception as e:
            logger.error(
                f'Failed to initialise catalog account details for share - {self.share.shareUri} '
                f'due to: {e}'
            )
            return None, None, None
        return None, None, None

```

**Review by dlpzx — COMMENTED** [review]
Small cosmetic changes. I like the new way of initializing the clients

**Comment by dlpzx** [comment]
Testing locally
- Testing datasets with catalog_account_id = dataset environment account
     - [ ] Sharing 2 tables + 1 folder --> succeeds and table can be queried from Athena
     - [ ] Revoking 2 tables + 1 folder --> succeeds
- Testing datasets with catalog_account_id != dataset environment account + WITHOUT proper tags +  catalog_account_id NOT linked as environment
     - [ ] Sharing 2 tables + 1 folder -->   raises error: Pivot role ... All tables appear as failed. Folders are shared.
     - [ ] Revoking 2 tables + 1 folder (first share then remove Environment/pivot role) --> raise error: Pivot Role.. All tables appear as failed. Folders are revoked.
- Testing datasets with catalog_account_id != dataset environment account + WITH proper tags +  catalog_account_id NOT linked as environment
     - [ ] Sharing 2 tables + 1 folder -->   raises error: Pivot role... All tables appear as failed. Folders are shared.
     - [ ] Revoking 2 tables + 1 folder (first share then remove Environment/pivot role) --> raise error: Pivot role... All tables appear as failed. Folders are revoked.
- Testing datasets with catalog_account_id != dataset environment account + WITHOUT proper tags +  catalog_account_id linked as environment
     - [ ] Sharing 2 tables + 1 folder -->  raises error: Tags ... All tables appear as failed. Folders are shared.
     - [ ] Revoking 2 tables + 1 folder (first share then remove tags) --> raise error: Tags... All tables appear as failed. Folders are revoked.
- Testing datasets with catalog_account_id != dataset environment account + WITH proper tags +  catalog_account_id linked as environment
     - [ ] Sharing 2 tables + 1 folder --> succeeds and table can be queried from Athena
     - [ ] Revoking 2 tables + 1 folder --> succeeds

File: backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py
Comment (TejasRGitHub) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:72]: ohh, I see. Thanks for the suggesion. I will update the PR accorindingly.  I was under the impression that `data_sharing_service.py` handles if I return false. but I see from the code it doesn't.

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (TejasRGitHub) [reply backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:510]: What benefit would that provide ? 

So are you suggesting from the point of view that when a catalog_dict is formulated the responsiblity of `_verify_catalog_ownership` remains only to check if pivot role is assumable and that the tags are present

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (TejasRGitHub) [reply backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:524]: Initially this function was more like what you have mentioned but in order to split it and make more generic this functions was split into two . I do not mind combining it into one in favor of reducing duplication. Do you think it would be beneficial to do it this way or keep it separate so that it can be used separately and have separate responsibilities ?

File: backend/dataall/base/aws/sts.py
Comment (TejasRGitHub) [reply backend/dataall/base/aws/sts.py:365]: Oh. Okay I see where the issue might be present. I will catch the exception as well. Thanks for pointing that out @dlpzx , @noah-paige

**Comment by noah-paige** [comment]
Reviewing code from a high level looks good with the latest edits - but will defer to @dlpzx more formal review and testing for this PR

File: backend/dataall/base/aws/sts.py
Comment (dlpzx) [reply backend/dataall/base/aws/sts.py]: Looks good and clean now, resolving

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:524]: I see now where it comes from. We can leave it separate in case we want to re-use get_catalog_account_details in the future

File: backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py:510]: It is more from a readability POV. My first thought when I saw the function was "wait are we verifying the ownership for the case environment account = catalog account?" But it is very minor, one can go to the function and see that we only verify when the accounts are different

**Comment by SofiaSazonova** [comment]
@dlpzx locally it looks fine. It's ready to be tested in AWS

**Comment by dlpzx** [comment]
Testing in AWS. To not repeat the exact same tests I will try some additional scenarios
- [X] CICD pipeline succeeds
- Testing datasets with catalog_account_id = dataset environment account (SAME ACCOUNT sharing and KMS imported Dataset)
     - [X] Sharing 2 tables + bucket --> succeeds and table can be queried from Athena, objects from bucket can be downloaded
     - [X] Revoking 2 tables + bucket --> succeeds
- Testing datasets with catalog_account_id != dataset environment account + WITHOUT proper tags +  catalog_account_id NOT linked as environment. For this test, I created one S3 bucket in the data producer account, then I registered the S3 location in Lake formation in another account (catalog account) with an IAM role with permissions in the data producer bucket policy. I created the database and tables in the catalog account, I created the schema of the tables and I was able to query the data in Athena in the catalog account. Then I granted permissions to this database and tables to the data producer account with Lake Formation. In the data producer account I accepted the RAM invitations and created a resource link database targeting the shared database. I can see the tables of the resource link database (the same ones as in the original database). Then I import the dataset into data.all using the original S3 bucket and the resource link database.
     - [ ] Sharing 2 tables + 1 folder -->   raises error: Pivot role ... All tables appear as failed. Folders are shared.
     - [ ] Revoking 2 tables + 1 folder (first share then remove Environment/pivot role) --> raise error: Pivot Role.. All tables appear as failed. Folders are revoked.
- Testing datasets with catalog_account_id != dataset environment account + WITH proper tags +  catalog_account_id NOT linked as environment
     - [ ] Sharing 2 tables + 1 folder -->   raises error: Pivot role... All tables appear as failed. Folders are shared.
     - [ ] Revoking 2 tables + 1 folder (first share then remove Environment/pivot role) --> raise error: Pivot role... All tables appear as failed. Folders are revoked.
- Testing datasets with catalog_account_id != dataset environment account + WITHOUT proper tags +  catalog_account_id linked as environment
     - [ ] Sharing 2 tables + 1 folder -->  raises error: Tags ... All tables appear as failed. Folders are shared.
     - [ ] Revoking 2 tables + 1 folder (first share then remove tags) --> raise error: Tags... All tables appear as failed. Folders are revoked.
- Testing datasets with catalog_account_id != dataset environment account + WITH proper tags +  catalog_account_id linked as environment
     - [ ] Sharing 2 tables + 1 folder --> succeeds and table can be queried from Athena
     - [ ] Revoking 2 tables + 1 folder --> succeeds

**Review by dlpzx — APPROVED** [review]
lgtm. I left a minor suggestion but it is optional

**Review by noah-paige — APPROVED** [review]
looks good!

**Comment by TejasRGitHub** [comment]
> lgtm. I left a minor suggestion but it is optional

I updated the code for the minor comment. I think it will help with readability and also goes along with the comments. Thanks   @dlpzx

**Pull request merged**
